hail/src/main/scala/is/hail/expr/ir/agg/AppendOnlyBTree.scala

```diff
@@ -201,30 +201,61 @@ class AppendOnlyBTree(kb: EmitClassBuilder[_], key: BTreeKey, region: Value[Regi
     getF.invokeCode(root, km, kv)
 
   def foreach(cb: EmitCodeBuilder)(visitor: (EmitCodeBuilder, Code[Long]) => Unit): Unit = {
-    val f = kb.genEmitMethod("btree_foreach", FastIndexedSeq[ParamType](typeInfo[Long]), typeInfo[Unit])
-    val node = f.getCodeParam[Long](1)
-    val i = f.newLocal[Int]("aobt_foreach_i")
+    val stackI = cb.newLocal[Int]("btree_foreach_stack_i", -1)
+    val nodeStack = cb.newLocal("btree_foreach_node_stack", Code.newArray[Long](const(128)))
+    val idxStack = cb.newLocal("btree_foreach_index_stack", Code.newArray[Int](const(128)))
+
+    def stackPush(node: Code[Long]) = {
+      cb.assign(stackI, stackI + 1)
+      cb += nodeStack.update(stackI, node)
+      cb += idxStack.update(stackI, -1)
+    }
+    def stackUpdateIdx(newIdx: Code[Int]) = {
+      cb += idxStack.update(stackI, newIdx)
+    }
+    def stackPop() = {
+      cb.assign(stackI, stackI - 1)
+    }
 
-    f.emitWithBuilder { cb =>
+    stackPush(root)
+    cb.whileLoop(stackI >= 0, { (Lstart) =>
+      val node = cb.newLocal("btree_foreach_node", nodeStack(stackI))
+      val idx = cb.newLocal("btree_foreach_idx", idxStack(stackI))
+      val Lend = CodeLabel()
+      val Lminus1 = CodeLabel()
+      val labels = Array.fill[CodeLabel](maxElements)(CodeLabel())
+      // this should probably be a switch, don't know how to make it one though
+      // furthermore, we should be able to do the lookups at runtime
+      // FIXME, clean this up once we have fixed arrays
```

Comment:
Not for this PR, but let's talk about the interface for this.  `cb.switch(i: Int, default: CodeLabel, cases: Array[CodeLabe])` should work for this case, and should straightforwardly follow from `Code.switch`.